### ISSUE TITLE
Practice for error messages

### ISSUE THREAD

**Issue by johnwikman** [miking-lang/miking#110]
(Based on [this comment](https://github.com/miking-lang/miking/pull/108#discussion_r447555654))

Many error messages do not provide details on how the error occurred, just what went wrong. For example with `mapLookup` in [stdlib/map.mc](https://github.com/miking-lang/miking/blob/develop/stdlib/map.mc#L18) it just states that no key was found, not _which_ key that wasn't found:
```ocaml
let mapLookup = lam eq. lam k. lam seq.
  optionMapOrElse (mapLookupOpt eq k seq)
                  (lam _. error "Element not found")
                  (lam x. x)
```

Instead we should probably help the user a bit more by specifying which key that wasn't found:
```ocaml
let mapLookup = lam eq. lam k. lam seq.
  optionGetOrElse (lam _. error (strJoin " " ["Element bound to key \"", k, "\" was not found"]))
                  (mapLookupOpt eq k seq)
```
Or by making an `errorFmt` function to simplify the error reporting:
```ocaml
let mapLookup = lam eq. lam k. lam seq.
  optionGetOrElse (lam _. errorFmt "Element bound to key \"$0\" was not found" [k])
                  (mapLookupOpt eq k seq)
```

The issue with these examples is that we assume that the `k` variable has a builtin way of converting it to a string, which might not always be the case. Currently without a static type system it is possible to get around this by assuming that it is a string and throwing an error if it is not, but then we produce a completely unrelated error message about the key not being a string.

One possible approach to resolve this would be to enforce that all types should have a `show` function. Though that would require that we more or less have a polymorphic function that in runtime converts the referenced value to a string. Is this an acceptable solution for MCore, or is there a better way to do error prints?

**Comment by dlunde** [comment]
I agree, we should of course provide more helpful error messages (at least in the future when things stabilize). I currently take a lazy approach when programming: make a simple static error message first, and then expand it as needed for debugging when the error actually occurs. For our current situation, I have found this to be good enough.
